src/dataflow/src/render/reduce.rs

```diff
@@ -144,94 +142,111 @@ where
-                            // case, rather than panic, though we surely want to see what is up.
-                            // XXX: This panic reports user-supplied data!
-                            panic!(
-                                "[customer-data] ReduceCollation found unexpected indexes:\n\tExpected:\t{:?}\n\tFound:\t{:?}\n\tFor:\t{:?}\n\tKey:{:?}\nRelationExpr:\n{}",
-                                (0..aggregates_len).collect::<Vec<_>>(),
-                                input.iter().map(|((p,_),_)| p).collect::<Vec<_>>(),
-                                aggregates_clone,
-                                key,
-                                relation_expr_clone.pretty(),
-                            );
-                        }
-                        row_packer.extend(key.iter());
-                        for ((_pos, val), cnt) in input.iter() {
-                            assert_eq!(*cnt, 1);
-                            row_packer.push(val.unpack_first());
-                        }
-                        output.push((row_packer.finish_and_reuse(), 1));
-                    }
-                });
-                (oks, err_input)
-            };
-            let index = (0..group_key.len()).collect::<Vec<_>>();
-            self.set_local_columns(relation_expr, &index[..], (oks, errs.arrange()));
+                    let is_accumulable = aggregates
+                        .iter()
+                        .map(|a| accumulable_hierarchical(&a.func).0)
+                        .collect::<Vec<_>>();
+                    differential_dataflow::collection::concatenate(scope, to_collect)
+                        .reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceCollation", {
+                            let mut row_packer = RowPacker::new();
+                            move |key, mut input, output| {
+                                // The inputs are pairs of an optional index and row to decode.
+
+                                // There can be at most one `None` index, and it indicates the accumulable aggregates.
+                                let new_row = Row::new(Vec::new());
+                                let mut accumulable = if (input[0].0).0 == None {
+                                    let iter = (input[0].0).1.iter();
+                                    input = &input[1..];
+                                    iter
+                                } else {
+                                    new_row.iter()
+                                };
+
+                                row_packer.extend(key.iter());
+                                for is_accum in is_accumulable.iter() {
+                                    if *is_accum {
+                                        row_packer.push(accumulable.next().unwrap());
+                                    } else {
+                                        let elem = input[0].0;
```

Comment:
what guarantees that these inputs will be in the right order? @frankmcsherry is it that `reduce_abelian` guarantees some ordering of inputs to this closure?

Reply:
It does provide that guarantee, yes!